src/compiler/ng-jest-compiler.spec.ts

```diff
@@ -1,34 +1,587 @@
+            expect(code).toContain(dedent`
+              let Wrapper = class Wrapper {
+                    constructor(y) {
+                        let ShouldBeProcessed = class ShouldBeProcessed {
+                            constructor(x) { }
+                        };
+                        ShouldBeProcessed.ctorParameters = () => [
+                            { type: ZoneToken }
+                        ];
+                        ShouldBeProcessed = tslib_1.__decorate([
+                            (0, core_1.Injectable)()
+                        ], ShouldBeProcessed);
+                    }
+              };
+              exports.Wrapper = Wrapper;
+            `);
+        });
+
+        // Angular is not concerned with type information for decorated class members. Instead,
+        // the type is omitted. This also helps with server side rendering as DOM globals which
+        // are used as types, do not load at runtime. https://github.com/angular/angular/issues/30586.
+        it('should downlevel Angular-decorated class member but not preserve type', () => {
+            const { code, diagnostics } = transformCjs(`
+               import {Input} from '@angular/core';
+               import {MyOtherClass} from './other-file';
+
+               export class MyDir {
+                 @Input() trigger: HTMLElement;
+                 @Input() fromOtherFile: MyOtherClass;
+               }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain(dedent`
+               MyDir.propDecorators = {
+                 trigger: [{ type: core_1.Input }],
+                 fromOtherFile: [{ type: core_1.Input }]
+               };
+            `);
+            expect(code).not.toContain('HTMLElement');
+            expect(code).not.toContain('MyOtherClass');
+        });
+
+        it.failing('should capture constructor type metadata with `emitDecoratorMetadata` enabled', () => {
+            // This test only passes when doing `import type {MyOtherClass} from './other-file';`
+            const { code, diagnostics } = transformCjs(
+                `
+                    import {Directive} from '@angular/core';
+                    import {MyOtherClass} from './other-file';
+
+                    @Directive()
+                    export class MyDir {
+                        constructor(other: MyOtherClass) {}
+                    }
+                `,
+                { emitDecoratorMetadata: true },
+            );
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain('const other_file_1 = require("./other-file");');
+            expect(code).toContain(dedent`
+                MyDir.ctorParameters = () => [
+                    { type: other_file_1.MyOtherClass }
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)(),
+                    tslib_1.__metadata("design:paramtypes", Object])
```

Comment:
My understanding of `emitDecoratorMetadata` is that it changes the output slightly based on whether or not the constructor parameter type is local or imported. It wants to emit `tslib_1.__metadata("design:paramtypes", MyOtherClass)`, but TS can't introspect the type of an imported symbol with `isolatedModules` and doesn't know `MyOtherClass` is a `class`. Instead, TS needs to generate a small JS snippet to determine this at runtime. See the output from [this example](https://www.typescriptlang.org/play/?experimentalDecorators=true&emitDecoratorMetadata=true&target=10&jsx=0&module=7&isolatedModules=true&isolatedDeclarations=true#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwGEcBbABzxFQwAoBKALnmo1gHMQNGpUBPW+ALwA+eADccWYAG4AUDID08+AAEMAZwC0WVqhxwZWMnozwA3gEkjMDCGABlEDFFYwIAL5IYJeAHIAdPJqjs6uPrIykFBqavAAMjhgUBAOTi4Ipm5yykRGqJQ0tDIgAB7k1vCR0fAAsjw55HlUZjLwrRV4ahgwyJh61C1tg6QwWKJQNvBwUMB4EDzwEAlJjPGJycFpADQDg63Do+MIUzOoc-CGZTbAjJaXtikhINuD-BkymUA). Is that the output you get which causes the test failure?

Interestingly, `tsc` seems to add this snippet whether or not `isolatedModules` is enabled. In theory I can see that TS wouldn't know `MyOtherClass` is a class as opposed to an interface and maybe needs to pick `Object` because it can't be certain `MyOtherClass` is valid in a value reference position.

Reply:
Yes the output from TS playground was exactly the output which caused the test failture. If we leave the output codes to be similar like TS playground, will that cause any issues when testing real application codes? 

Is `emitDecoratorMetadata: true` still important to Angular apps these days? I remember somewhere I saw a commit from Angular which allows to disable `emitDecoratorMetadata` and application still works.

Reply:
I think that emitted JS is probably fine? Hard to be 100% sure unfortunately.

`ng new` does not enable `emitDecoratorMetadata` in the scaffolded `tsconfig.json` files, so I suspect it's not required, but I wouldn't be surprised if there are other effects from enabling it. I can ask around and see if anyone has more context.